packages/server/lib/browsers/geckodriver/index.ts

```diff
@@ -0,0 +1,122 @@
+      allowHosts?: string[]
+      allowOrigins?: string[]
+      binary?: string
+      connectExisting?: boolean
+      host?: string
+      jsdebugger?: boolean
+      log?: 'fatal' | 'error' | 'warn' | 'info' | 'config' | 'debug' | 'trace'
+      logNoTruncate?: boolean
+      marionetteHost?: string
+      marionettePort?: number
+      port?: number
+      websocketPort?: number
+      profileRoot?: string
+      geckoDriverVersion?: string
+      customGeckoDriverPath?: string
+      cacheDir?: string
+      spawnOpts: SpawnOpt
+    }) => Promise<ChildProcess>
+    download: (geckodriverVersion?: string, cacheDir?: string) => Promise<string>
+  } = () => {
+      /**
+       * NOTE: geckodriver is an ESM package and does not play well with mksnapshot.
+       * Requiring the package in this way, dynamically, will
+       * make it undiscoverable by mksnapshot
+       */
+      return require(require.resolve(geckoDriverPackageName, { paths: [__dirname] }))
+    }
+
+  /**
+   * creates an instance of the GeckoDriver server. This is needed to start WebDriver
+   * @param {StartGeckoDriverArgs} opts - arguments needed to start GeckoDriver
+   * @returns {ChildProcess} - the child process in which the geckodriver is running
+   */
+  static async create (opts: StartGeckoDriverArgs, timeout: number = 5000): Promise<ChildProcess> {
+    debug('no geckodriver instance exists. starting geckodriver...')
+
+    let geckoDriverChildProcess: ChildProcess | null = null
+
+    try {
+      const { start: startGeckoDriver } = GeckoDriver.getGeckoDriverPackage()
+
+      geckoDriverChildProcess = await startGeckoDriver({
+        host: opts.host,
+        port: opts.port,
+        marionetteHost: opts.marionetteHost,
+        marionettePort: opts.marionettePort,
+        websocketPort: opts.webdriverBidiPort,
+        profileRoot: opts.profilePath,
+        binary: opts.binaryPath,
+        jsdebugger: debugModule.enabled(GECKODRIVER_DEBUG_NAMESPACE_VERBOSE) || false,
+        log: debugModule.enabled(GECKODRIVER_DEBUG_NAMESPACE_VERBOSE) ? 'debug' : 'error',
+        logNoTruncate: debugModule.enabled(GECKODRIVER_DEBUG_NAMESPACE_VERBOSE),
+        spawnOpts: opts.spawnOpts || {},
+      })
+
+      // using a require statement to make this easier to test with mocha/mockery
+      const waitPort = require('wait-port')
+
+      await Bluebird.resolve(waitPort({
+        port: opts.port,
+        // add 1 second to the timeout so the timeout throws first if the limit is reached
+        timeout: timeout + 1000,
+        output: debugModule.enabled(GECKODRIVER_DEBUG_NAMESPACE_VERBOSE) ? 'dots' : 'silent',
+      })).timeout(timeout)
+
+      debug('geckodriver started!')
+
+      // For whatever reason, we NEED to bind to stderr/stdout in order
+      // for the geckodriver process not to hang, even though the event effectively
+      // isn't doing anything without debug logs enabled.
+      geckoDriverChildProcess.stdout?.on('data', (buf) => {
```

Comment:
@whimboo do you have any idea why the above comment would happen? @mschile and I tried gating the `stdout`/`stderr` behind an if statement and it looks to cause the process to almost freeze (is still running but appears stuck). We tested this on Firefox 125/130 with geckodriver `0.35.0`. Is it possible the driver is timing out or something like that?

Reply:
No I don't know why that happens. If you may have the time and could create a minimal stand-alone testcase to reproduce this behavior it would be great. We haven't heard about it yet. 

One thing which could happen is if a pipe was setup for stdout/stderr but nothing is read from it. That means that geckodriver will run into an overflow slowing down massively first. We had something similar in the past for Puppeteer when logging was enabled but the pipe was not read. It was fixed via https://github.com/puppeteer/puppeteer/pull/7644.